### PR TITLE
Multicall

### PR DESCRIPTION
**Pull request by bowd**
> See [Readme](https://github.com/Alethio/web3-go/blob/c0772447e80d2dd059836dc187dadc96dfa01e72/multicall/README.md)

#### TL;DR

Batch several `view` contract calls into a single one using the [multicall](https://github.com/bowd/multicall/blob/master/src/Multicall.sol) contract I forked from Maker.
This means we get to read state in a single EVM trip which _might*_ be faster but will ensure read atomicity when reading at the tip of the chain. 

> (might*) Haven't properly tested empirically, will do that when testing its usage with our internal tools.

### CODE REVIEW

File: multicall/multicall_test.go [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+		0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0,
+		0x0, 0x0, 0x0, 0xc}
+	assert.Equal(t, expectedArgTypes, vc.argumentTypes())
+	callData, err := vc.callData()
+	assert.Nil(t, err)
+	assert.Equal(t, expectedCallData, callData)
+}
+
+func ExampleViwCall() {
+	eth, err := getETH("https://mainnet.infura.io/v3/17ed7fe26d014e5b9be7dfff5368c69d")
```
Comment (tzapu) [inline multicall/multicall_test.go]: should not have credentials in code

Comment (bowd) [reply multicall/multicall_test.go]: Should be smarter, indeed

**Comment by lacasian** [comment]
Don't want to upset anyone here but I'm wondering if web3-go is the right place to implement this method considering it is dependent on having the contract deployed. 
I'm just curious if this was taken into consideration.

**Comment by bowd** [comment]
@kwix good question, but I don't really have an answer.

 It does come with presets to the deployed contract on some nets (we can deploy it on most public nets) so people have to deploy a contract in order to use this only on private chains. The one I forked is up with sourcecode verified. At the same time, I didn't want to reimplement the ETH logic so even if we extract it to a separate repo I would make it dependent on `web3-go` but I guess that's fine. So I'm good either way. I added it here because I saw this as a compendium of ways to integrate with web3 from go, but I'm happy to extract if we want to make the repo more focused.

**Comment by tzapu** [comment]
not sure what happend with my comment above, can t add so  i ll repost here:

"this adds geth as a dependency, which is a heavy dep to have, with C bindings and stuff like that. It also adds issues to other projects using geth in parallel due to versioning.
Had a few problems in the past because of this, they might be less now as i think they finally moved to go modules, but this is still outside the scope of web3-go imho.
can you think of another way of doing it? separate package?"

a separate pkg dependent of web3-go and go-ethereum seems preferable to me

**Comment by bowd** [comment]
Done! https://github.com/Alethio/multicall-go
Thanks for helping me see the forest despite the trees 👍

**Pull request closed without merging**